[WIP] Making `Camembert` independent from Roberta

Mustapha-AJEGHRIR opened the pull request:
# What does this PR do?

related to  #19303 

Making the Camembert model (pytorch version) independent from Roberta

I have changed all the classes of camembert by copy paste from Roberta. I have done the litte necessary changes for every thing to work.

I'm still wondering how to change blocks like those, I have to  add a specific checkpoint for Camembert ? :
```python
    @add_code_sample_docstrings(
        processor_class=_TOKENIZER_FOR_DOC,
        checkpoint="deepset/roberta-base-squad2",
        output_type=QuestionAnsweringModelOutput,
        config_class=_CONFIG_FOR_DOC,
        expected_output="' puppet'",
        expected_loss=0.86,
    )
```
For testing, the following test works well :
```bash
$ RUN_SLOW=1 pytest tests/models/camembert/test_modeling_camembert.py
```

However, I have noticed the tests are only for the `CamembertModel` but not other classes like `CamembertForCausalLM` ...

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@sgugger related to  #19303

Commit:
Tests passed for modeling_camembert

Commit:
Tests passed for modeling_camembert

Commit:
fixing the name 'roberta'

Commit:
fixing the name 'roberta'

Mustapha-AJEGHRIR commented:
In the `CamembertPreTrainedModel` I left `base_model_prefix = "roberta"` to be equal to "roberta". If I change it into "camembert" the test fails. This is normal ?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19312). All of your documentation changes will be reflected on that endpoint.

Commit:
Code quality

Commit:
Code quality

patrickvonplaten committed:
[Docs] Fix link (#19313)

Rocketknight1 committed:
Fix for sequence regression fit() in TF (#19316)

Co-authored-by: Your Name <you@example.com>

IMvision12 committed:
Added Type hints for LED TF (#19315)

* Update modeling_tf_led.py

* Update modeling_tf_led.py

debjit-bw committed:
Added type hints for TF: rag model (#19284)

* Added type hints for TF: rag model

* TFModelInputType added in place of TF.Tensor

* reformatting by black

gouqi666 committed:
alter retrived to retrieved (#18863)

oscard0m committed:
ci(stale.yml): upgrade actions/setup-python to v4 (#19281)

oscard0m committed:
ci(workflows): update actions/checkout to v3 (#19280)

in stale.yml

daspartho committed:
wrap forward passes with torch.no_grad() (#19279)

daspartho committed:
wrap forward passes with torch.no_grad() (#19278)

daspartho committed:
wrap forward passes with torch.no_grad() (#19274)

daspartho committed:
wrap forward passes with torch.no_grad() (#19273)

arnaudstiegler committed:
Removing BertConfig inheritance from LayoutLMConfig (#19307)

* removing BertConfig inheritance

* fix missing arguments

Sushrut1101 committed:
docker-build: Update actions/checkout to v3 (#19288)

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return position_ids.unsqueeze(0).expand(input_shape)
+
+
+# Copied from transformers.models.bert.modeling_bert.BertSelfAttention with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertSelfAttention with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return outputs
+
+
+# Copied from transformers.models.bert.modeling_bert.BertSelfOutput with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertSelfOutput with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return hidden_states
+
+
+# Copied from transformers.models.bert.modeling_bert.BertAttention with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertAttention with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return outputs
+
+
+# Copied from transformers.models.bert.modeling_bert.BertIntermediate with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertIntermediate with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return hidden_states
+
+
+# Copied from transformers.models.bert.modeling_bert.BertOutput with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertOutput with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return hidden_states
+
+
+# Copied from transformers.models.bert.modeling_bert.BertLayer with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertLayer with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        return layer_output
+
+
+# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        )
+
+
+# Copied from transformers.models.bert.modeling_bert.BertPooler with Bert->Roberta->Camembert
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertPooler with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+    """
+
+    config_class = CamembertConfig
+    base_model_prefix = "roberta"
```
Yes, this needs to stay as is to be compatible with the existing weights (this is the name the body of the model will have in the models with heads).

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
-    config_class = CamembertConfig
+    _keys_to_ignore_on_load_missing = [r"position_ids"]
+
+    # Copied from transformers.models.bert.modeling_bert.BertModel.__init__ with Bert->Roberta->Camembert
```
```suggestion
    # Copied from transformers.models.bert.modeling_bert.BertModel.__init__ with Bert->Camembert
```

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -55,104 +75,1509 @@
+        self.bias = self.decoder.bias
+
+
+# Copied from transformers.models.roberta.modeling_reoberta.RobertaModel with Roberta->Camembert
```
Not sure if this copied from will work since there are two others for methods. If the tests are happy, we can keep it.

Mustapha-AJEGHRIR replied:
It didn't work so I removed it

sgugger reviewed (commented):
Thansk a lot for starting to work on this! I left a couple of comments and there are a few inconsistencies found by the copy checker (if you run locally `make repo-consistency` you should see the errors).

You'll also need a couple of extra new lines to make flake8 happy (run `make quality` locally to see where it complains).

Commit:
Code quality fake8 and repo-consistency

Commit:
Code quality fake8 and repo-consistency

Mustapha-AJEGHRIR commented:
Thanks a lot for your answers @sgugger. I took some time to understand the concepts of `flake8` and `repo-consistency-

Mustapha-AJEGHRIR commented:
I have done one thing without being very sure about it (that was the only way to make `make fixup` happy). I added the `CamembertPreTrainedModel` class into the `__init__` and `dummy_pt_objects` (as in the PR). I don't know if this is the correct way to do it ?

Commit:
Adding CamembertPreTrainedModel to __init__

Commit:
Adding CamembertPreTrainedModel to __init__

younesbelkada committed:
Add `BloomForQuestionAnswering` (#19310)

* add bloom for question answering

- attempt to add Bloom for question answering
- adapted from `GPTJForQuestionAnswering`
- Fixed `num_labels` to `2` for common tests
- Added a bit of docstring
- All common tests pass

* Update src/transformers/models/bloom/modeling_bloom.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* revert changes related to `num_labels`

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger reviewed (approved):
For the dummy objects. `make fix-copies` would have done it automatically. The quality tests all seem to be green, so you did well :-)

Let's double-check all tests pass but we should be ready to merge (be sure to pass from draft to ready for review!)

sgugger commented on src/transformers/models/camembert/modeling_camembert.py:
```diff
@@ -56,30 +76,948 @@
+                "bi-directional self-attention."
+            )
+
+        self.camembert = CamembertModel(config, add_pooling_layer=False)
```
```suggestion
        self.roberta = CamembertModel(config, add_pooling_layer=False)
```
This one needs to stay at `roberta` (otherwise pretrained weights won't work anymore). It's the same for all other models with head.

sgugger reviewed (commented):
As pointed in the comment above, we need to keep the `self.roberta` for all models with head (not move it to `self.camembert`) otherwise the checkpoint (which use the "roberta" key) won't load in the model anymore.

SSamDav committed:
Clamping hidden state values to allow FP16 (#19229)

* Clamping hidden state values to allow FP16

* Reformating

* Adding missing if condition

* Update src/transformers/models/longt5/modeling_longt5.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* Update src/transformers/models/longt5/modeling_longt5.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* Update src/transformers/models/longt5/modeling_longt5.py

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* Formating file

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

Mustapha-AJEGHRIR committed:
Update src/transformers/models/camembert/modeling_camembert.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Commit:
leave self.roberta in camembert

Commit:
leave self.roberta in camembert

Commit:
Merge branch 'camembert-independent-from-roberta' of https://github.com/Mustapha-AJEGHRIR/transformers into camembert-independent-from-roberta

E-Aho committed:
Remove interdependency from OpenAI tokenizer (#19327)

* Remove interdependency from OpenAI tokenizer

* Adjust import order for linter

D3xter1922 committed:
removing XLMConfig inheritance from FlaubertConfig (#19326)

* removing XLMConfig inheritance from FlaubertConfig

* removing XLMConfig inheritance from FlaubertConfig

* Fixed styling issue

* Update configuration_flaubert.py

Co-authored-by: Druhin Abrol <druhinabrol@192.168.1.6>

Mustapha-AJEGHRIR commented:
I don't understand why some torch tests failed just by replacing `self.camembert` with `self.roberta`.

Commit:
Merge branch 'camembert-independent-from-roberta' of https://github.com/Mustapha-AJEGHRIR/transformers into camembert-independent-from-roberta

Mustapha-AJEGHRIR commented:
Hi @sgugger, I have done git rebase from the huggingface:main, this brought all the other commit over my branch, this might be very difficult to read. If you want I can close this pull request and open a clean one ?

sgugger commented:
Arg, yes we would need a clean PR: even if I can see the changes are not related, it will mess up authorship of this commit when we merge your PR (it will show all the other random commits of your rebase) and if somehow this PR introduces a bug, when we come look at it later, it will be hard to see what changed the error.

Tip: for GitHub you need to force-push after a rebase when your PR is already open (with `git push -u origin branch --force`)
